analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs

```diff
@@ -206,87 +182,44 @@ private static void CompareActualToExpected(string languageVersion, Diagnostic[]
-            VerifyIssue(languageVersion, expectedIssues, issueFilter, location, message, extraInfo, true, null);
-
-        private static void VerifySecondaryIssue(string languageVersion, IList<IssueLocation> expectedIssues, Func<IssueLocation, bool> issueFilter,
-            Location location, string message, string issueId) =>
-            VerifyIssue(languageVersion, expectedIssues, issueFilter, location, message, null, false, issueId);
-
-        private static string VerifyIssue(string languageVersion,
-                                          ICollection<IssueLocation> expectedIssues,
-                                          Func<IssueLocation, bool> issueFilter,
-                                          Location location,
-                                          string message,
-                                          string extraInfo,
-                                          bool isPrimary,
-                                          string primaryIssueId)
+        private static IEnumerable<IssueLocationPair> MatchPairs(CompilationIssues actualIssues, CompilationIssues expectedIssues)
         {
-            var lineNumber = location.GetLineNumberToReport();
-            var expectedIssue = expectedIssues
-                .Where(issueFilter)
-                .OrderBy(x => x.Start == null ? 0 : Math.Abs(location.GetLineSpan().StartLinePosition.Character - x.Start.Value))
-                .ThenBy(x => x.Message == message ? 0 : 1)  // Prefer issue with explicit matching message when possible. There can be more with different messages
-                .FirstOrDefault(issue => issue.LineNumber == lineNumber);
-            var issueType = IssueType(isPrimary);
-
-            if (expectedIssue == null)
+            var ret = new List<IssueLocationPair>();
+            foreach (var key in actualIssues.UniqueKeys())
             {
-                var issueId = primaryIssueId == null ? "" : $" [{primaryIssueId}]";
-                var seeOutputMessage = $"{Environment.NewLine}See output to see all actual diagnostics raised on the file";
-                var lineSpan = location.GetLineSpan().Span.ToString();
-                var exceptionMessage = string.IsNullOrEmpty(extraInfo)
-                    ? $"{languageVersion}: Unexpected {issueType} issue{issueId} on line {lineNumber}, span {lineSpan} with message '{message}'.{seeOutputMessage}"
-                    : extraInfo;
-                throw new UnexpectedDiagnosticException(location, exceptionMessage);
+                ret.AddRange(BestMatch(actualIssues.Remove(key), expectedIssues.Remove(key)));
             }
+            ret.AddRange(expectedIssues.Select(x => new IssueLocationPair(null, x)));
+            return ret;
+        }
 
-            if (expectedIssue.Message != null && expectedIssue.Message != message)
+        private static IEnumerable<IssueLocationPair> BestMatch(List<IssueLocation> actualIssues, List<IssueLocation> expectedIssues)
```

Comment:
I don't like the name `BestMatch` that much, since from the caller's perspective, what it does is return the imperfect matches.

I think we should call it something like `FindMismatches`.